File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.Collections;
import java.util.PriorityQueue;
import java.util.*;

class Solution {
    public static int[] solution(String[] operations) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());

        for (String operation : operations) {
            String[] temp = operation.split(" ");

            if (temp[0].equals("I")) {
                minHeap.add(Integer.parseInt(temp[1]));
                maxHeap.add(Integer.parseInt(temp[1]));
            } else if(!maxHeap.isEmpty() && temp[0].equals("D") && temp[1].equals("1")) {
                Integer max = maxHeap.poll();
                minHeap.remove(max);
            } else if(!minHeap.isEmpty() && temp[0].equals("D") && temp[1].equals("-1")) {
                Integer min = minHeap.poll();
                maxHeap.remove(min);
            }
        }

        int[] answer = new int[2];

        if (maxHeap.isEmpty() && minHeap.isEmpty()) {
            answer[0] = 0;
            answer[1] = 0;
        } else {
            answer[0] = maxHeap.poll();
            answer[1] = minHeap.poll();
        }

        return answer;
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.kt
import java.util.PriorityQueue

class Solution {
    fun solution(operations: Array<String>): IntArray {
        val answer = IntArray(size = 2)

        val mainHeap = PriorityQueue<Int>(reverseOrder())
        val subHeap = PriorityQueue<Int>()

        for (i: Int in operations.indices) {
            val split: List<String> = operations[i].split(" ")

            when (split[0]) {
                // 삽입
                "I" -> {
                    mainHeap.add(split[1].toInt())
                }

                // 제거
                "D" -> when (split[1]) {
                    // 최대값을 제거한다
                    "1" -> {
                        if (mainHeap.isNotEmpty()) {
                            mainHeap.poll()
                        }
                    }

                    // 최소값을 제거한다.
                    "-1" -> {
                        while (mainHeap.isNotEmpty()) {
                            subHeap.add(mainHeap.poll())
                        }

                        if (subHeap.isNotEmpty()) {
                            subHeap.poll()
                        }

                        // 힙 갱신
                        while (subHeap.isNotEmpty()) {
                            mainHeap.add(subHeap.poll())
                        }
                    }
                }
            }
        }

        if (mainHeap.isEmpty()) {
            answer[0] = 0
            answer[1] = 0
        } else {
            answer[0] = mainHeap.peek()

            while (mainHeap.isNotEmpty()) {
                subHeap.add(mainHeap.poll())
            }

            if (subHeap.isNotEmpty()) {
                answer[1] = subHeap.peek()
            }
        }

        return answer
    }
}